refactor(w3dview): Migrate from MEMBER_ADD/MEMBER_RELEASE macros to RefCountPtr<T>

bobtista opened the pull request:
* Merge after #1784

Replaces custom reference counting macros (MEMBER_ADD, MEMBER_RELEASE, SAFE_ADD_REF, SAFE_RELEASE_REF) with the type-safe RefCountPtr<T> smart pointer from the core library. 

Member Variables Migrated to RefCountPtr<T>:

- AssetInfoClass::m_pRenderObj
- ScreenCursorClass::m_pTexture, m_pVertMaterial
- PlaySoundDialogClass::SoundObj
- SpherePropertySheetClass::m_RenderObj
- RingPropertySheetClass::m_RenderObj
- CW3DViewDoc::m_pCRenderObj, m_pCAnimation, m_pCAnimCombo, m_pCursor, m_pCursorScene
- CGraphicView::m_pCamera, m_pLightMesh

Local Variable Usage:

- Replaced MEMBER_ADD/MEMBER_RELEASE with REF_PTR_SET/REF_PTR_RELEASE for local raw pointers in 15 files

Removed Macros (Utils.h):

- MEMBER_ADD
- MEMBER_RELEASE
- SAFE_ADD_REF
- SAFE_RELEASE_REF

Notes:

- Used Create_AddRef() when taking ownership of an existing pointer
- Used Create_NoAddRef() when wrapping factory methods that return ref=1
- Added .Peek() calls when passing to functions expecting raw pointers
- Updated getter methods to return .Peek() instead of raw RefCountPtr
- Preserved all original logic and formatting

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -284,8 +280,6 @@ CW3DViewDoc::OnNewDocument (void)
 			// Free the currently displayed object
 			SAFE_DELETE (m_pCAnimCombo);
-			MEMBER_RELEASE (m_pCAnimation);
-			MEMBER_RELEASE (m_pCRenderObj);
```
m_pCAnimation = NULL
m_pCRenderObj = NULL

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -284,8 +280,6 @@ CW3DViewDoc::OnNewDocument (void)
     {
```
The condition above is obsoleted.

bobtista commented on Core/Tools/W3DView/W3DViewDoc.h:
```diff
@@ -291,18 +292,18 @@ class CW3DViewDoc : public CDocument
-	HAnimClass *			m_pCAnimation;
+	RefCountPtr<HAnimClass>			m_pCAnimation;
 	HAnimComboClass *		m_pCAnimCombo;
 	LightClass *			m_pCSceneLight;
```
Just focusing on keeping this change small and simple, the other ones are a bit more complicated to me

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -610,24 +603,22 @@ CW3DViewDoc::Display_Emitter
+			// Remove this object from the scene
+			Remove_Object_From_Scene (m_pCRenderObj.Peek());
+		}
+		m_pCRenderObj.Clear();
```
This can be moved into scope above.

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -669,17 +660,14 @@ CW3DViewDoc::DisplayObject
-					m_pCRenderObj = NULL;
+					Remove_Object_From_Scene (m_pCRenderObj.Peek());
 			  }
+			  m_pCRenderObj.Clear();
```
This can be moved into scope above

bobtista commented:
> How about we first make a change where we just remove the duplicate Macros, without adding RefCountPtr? Then this change will be simpler? And then make small changes to introduce RefCountPtr slowly? I am afraid this will be a big review here otherwise given the implications. One mistake with ref counting and the game will crash.

Ok, check out 1784. Shall we continue this PR after rebasing and have it focus just on just W3DView, then additional RefCountPtr migrations can have separate PRs?

xezon commented on Core/Tools/W3DView/AssetInfo.cpp:
```diff
@@ -52,9 +52,8 @@ AssetInfoClass::Initialize (void)
-		if (prender_obj)
-			prender_obj->Add_Ref();
+		RenderObjClass *prender_obj = m_pRenderObj.Peek();
+		if (prender_obj) prender_obj->Add_Ref();
```
new line

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -171,15 +170,16 @@ CW3DViewDoc::CleanupResources (void)
 
 	if (m_pCursor != NULL) {
 		m_pCursor->Remove ();
+		m_pCursor.Clear();
```
Is this correct? It did not do that before.

xezon commented on Core/Tools/W3DView/EmitterInstanceList.cpp:
```diff
@@ -33,6 +33,7 @@
 
 #include "StdAfx.h"
 #include "EmitterInstanceList.h"
+#include "refcount.h"
```
This should not be included

xezon commented:
This still needs work.

xezon commented on Core/Tools/W3DView/PlaySoundDialog.cpp:
```diff
@@ -100,7 +99,6 @@ void
 PlaySoundDialogClass::OnCancel (void)
 {
 	SoundObj->Stop ();
-	REF_PTR_RELEASE (SoundObj);
```
`SoundObj.Clear()`, otherwise this does not release here as did before.

xezon commented on Core/Tools/W3DView/RingPropertySheet.cpp:
```diff
@@ -201,17 +198,17 @@ RingPropertySheetClass::Add_Object_To_Viewer (void)
-		doc->Reload_Displayed_Object ();
-		m_LastSavedName = m_RenderObj->Get_Name ();
-		REF_PTR_SET (m_RenderObj, (RingRenderObjClass *)doc->GetDisplayedObject ());
+	//
```
Indentation has changed here.

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -200,14 +200,6 @@ CW3DViewDoc::CleanupResources (void)
-    if (m_pCBackObjectScene)
-    {
-        // Free the scene object
-        m_pCBackObjectScene->Release_Ref ();
```
`m_pCBackObjectScene.Clear()` missing?

xezon replied:
Something is missing here. Code was removed with no substitute.

bobtista replied:
Oh this?
On main, CleanupResources() releases m_pCBackObjectScene twice:
- lines 157–167 — the real one: removes m_pCBackgroundObject from the scene, then Release_Ref() + = nullptr
- lines 202–206 — a second if (m_pCBackObjectScene) { Release_Ref(); = nullptr; }, which can never fire because the first block already nulled it

The PR deletes only the second (unreachable) block.

TLDR The deleted block was an unreachable duplicate - the actual release remains earlier in CleanupResources(). 

Do you want me to:
1. Restore the eight lines
Or 
2. Migrate m_pCBackObjectScene to RefCountPtr<SceneClass> - then the first block becomes .Clear(), the duplicate genuinely has a substitute

xezon replied:
Ok I now see it is a duplicate.

Why did you choose to not upgrade some of the raw pointers to `RefCountPtr` in W3DView?

bobtista replied:
Just keeping the change scoped smaller. Happy to convert them here or in a follow up PR

xezon replied:
How many more are there in W3DView?

bobtista replied:
There are 16 remaining owning raw-pointer storage declarations in W3DView: 14 ordinary members and 2 containers. 
Eight are in CW3DViewDoc: m_pCScene, m_pC2DScene, m_pCBackObjectScene, m_pCBackgroundObject, m_pCSceneLight, m_pCBackgroundBMP, m_pC2DCamera, and m_pCBackObjectCamera.  

Six scalar members are elsewhere, plus AdvancedAnimSheet::Anims[128] and EmitterInstanceListClass::m_List. The two containers need more involved changes, particularly Anims, which is sorted with qsort and exposed as HAnimClass **. 

I can convert the eight CW3DViewDoc members here and leave the unrelated members/containers for follow-ups, sound good?

xezon replied:
Yes

bobtista replied:
Done

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -395,12 +380,12 @@ CW3DViewDoc::InitScene (void)
-		ASSERT (m_pCursorScene);
+	// Instantiate a new 2D cursor scene
+	m_pCursorScene = RefCountPtr<SceneClass>::Create_NoAddRef(new SimpleSceneClass);
+	ASSERT (m_pCursorScene.Peek());
```
Indentation has changed.

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -669,16 +652,13 @@ CW3DViewDoc::DisplayObject
     {
         // Lose the animation
 		  SAFE_DELETE (m_pCAnimCombo);
-		  REF_PTR_RELEASE (m_pCAnimation);
```
`m_pCAnimation.Clear()` is missing?

xezon reviewed (commented):
This refactor needs another pass. Something went wrong the indentation of code. And behavior has changed.

bobtista commented:
Ready for re-review. Rebased onto latest main and it's behavior-consistent with the original. 

Note - these fixes can either be added here or to separate PRs if we want:

1. m_pCursor.Clear() in CleanupResources - pulled this back out so the cursor is only released when the document is destroyed (as before, now via the RefCountPtr destructor). Clearing it in CleanupResources would release/recreate the cursor on every cleanup (e.g. File→New), which is a behavior change.
2. Leak in DisplayObject (ghost-object branch) - there's a local RenderObjClass *m_pCRenderObj that shadows the member; it's Add_Ref()'d and added to the scene but never released, so the local's reference leaks. Left as-is here to keep behavior identical.
3. GraphicView::OnDestroy - kept the explicit m_pCamera/m_pLightMesh release. With RefCountPtr these could release via the destructor instead, but that shifts release timing off WM_DESTROY, so I left the explicit clears.

xezon commented on Core/Tools/W3DView/AssetInfo.h:
```diff
@@ -58,16 +58,15 @@ class AssetInfoClass
 			  m_AssetType (type),
 			  m_dwUserData (user_data),
-			  m_pRenderObj (nullptr)			{ REF_PTR_SET (m_pRenderObj, prender_obj); Initialize (); }
+			  m_pRenderObj (RefCountPtr<RenderObjClass>::Create_AddRef(prender_obj))			{ Initialize (); }
```
This is quite verbose. I created a simplification with https://github.com/TheSuperHackers/GeneralsGameCode/pull/3058 that allows to write it simpler afterwards.

xezon commented on Core/Tools/W3DView/SpherePropertySheet.cpp:
```diff
@@ -68,10 +67,9 @@ SpherePropertySheetClass::SpherePropertySheetClass
+	:	CPropertySheet(pszCaption, pParentWnd, iSelectPage),
+		m_RenderObj (RefCountPtr<SphereRenderObjClass>::Create_AddRef(sphere))
 {
-	REF_PTR_SET (m_RenderObj, sphere);
```
After https://github.com/TheSuperHackers/GeneralsGameCode/pull/3058 you could also do `m_RenderObj.Assign_AddRef(sphere)`. Either way is fine.

xezon commented on Core/Tools/W3DView/ScreenCursor.cpp:
```diff
@@ -149,7 +141,7 @@ ScreenCursorClass::Initialize ()
 ScreenCursorClass::Set_Texture (TextureClass *texture)
 {
-	REF_PTR_SET (m_pTexture, texture);
+	m_pTexture = RefCountPtr<TextureClass>::Create_AddRef(texture);
```
Do Assign_AddRef later

xezon commented on Core/Tools/W3DView/ScreenCursor.cpp:
```diff
@@ -101,10 +95,8 @@ ScreenCursorClass::~ScreenCursorClass ()
-
 	// Create default vertex material
-	m_pVertMaterial = NEW_REF( VertexMaterialClass, ());
+	m_pVertMaterial = RefCountPtr<VertexMaterialClass>::Create_NoAddRef(NEW_REF( VertexMaterialClass, ()));
```
Do Assign_AddRef later

xezon commented on Core/Tools/W3DView/RingPropertySheet.cpp:
```diff
@@ -270,12 +267,12 @@ RingPropertySheetClass::Initialize ()
 RingPropertySheetClass::Create_New_Object ()
 {
-	m_RenderObj = new RingRenderObjClass;
+	m_RenderObj = RefCountPtr<RingRenderObjClass>::Create_NoAddRef(new RingRenderObjClass);
```
Do Assign_NoAddRef later

xezon commented on Core/Tools/W3DView/RingPropertySheet.cpp:
```diff
@@ -201,14 +198,14 @@ RingPropertySheetClass::Add_Object_To_Viewer ()
 		doc->Reload_Displayed_Object ();
 		m_LastSavedName = m_RenderObj->Get_Name ();
-		REF_PTR_SET (m_RenderObj, (RingRenderObjClass *)doc->GetDisplayedObject ());
+		m_RenderObj = RefCountPtr<RingRenderObjClass>::Create_AddRef((RingRenderObjClass *)doc->GetDisplayedObject ());
```
Do Assign_AddRef later

xezon commented on Core/Tools/W3DView/PlaySoundDialog.cpp:
```diff
@@ -120,7 +119,7 @@ PlaySoundDialogClass::OnInitDialog ()
 	//	Create the sound effect so we can play it
 	//
-	SoundObj = WWAudioClass::Get_Instance ()->Create_Sound_Effect (Filename);
+	SoundObj = RefCountPtr<AudibleSoundClass>::Create_NoAddRef(WWAudioClass::Get_Instance ()->Create_Sound_Effect (Filename));
```
Do Assign_NoAddRef later

xezon commented on Core/Tools/W3DView/GraphicView.cpp:
```diff
@@ -240,7 +238,7 @@ CGraphicView::InitializeGraphicView ()
 		WW3DAssetManager::Get_Instance()->Load_3D_Assets (light_mesh_file);
 
-		m_pLightMesh = WW3DAssetManager::Get_Instance()->Create_Render_Obj ("LIGHT");
+		m_pLightMesh = RefCountPtr<RenderObjClass>::Create_NoAddRef(WW3DAssetManager::Get_Instance()->Create_Render_Obj ("LIGHT"));
```
Do Assign_NoAddRef later

xezon commented on Core/Tools/W3DView/GraphicView.cpp:
```diff
@@ -212,7 +210,7 @@ CGraphicView::InitializeGraphicView ()
     {
         // Instantiate a new camera class
-	    m_pCamera = new CameraClass ();
+	    m_pCamera = RefCountPtr<CameraClass>::Create_NoAddRef(new CameraClass ());
```
Do Assign_NoAddRef later, many times in this change.

xezon commented on Core/Tools/W3DView/AssetInfo.h:
```diff
@@ -83,14 +82,14 @@ class AssetInfoClass
 		const CString &	Get_User_String () const				{ return m_UserString; }
-		RenderObjClass *	Get_Render_Obj () const				{ if (m_pRenderObj) m_pRenderObj->Add_Ref(); return m_pRenderObj; }
-		RenderObjClass *	Peek_Render_Obj () const				{ return m_pRenderObj; }
+		RenderObjClass *	Get_Render_Obj () const				{ RenderObjClass *ptr = m_pRenderObj.Peek(); if (ptr) ptr->Add_Ref(); return ptr; }
```
Ideally in the future functions like this would also return a `RefCountPtr<T>`, but is beyond the scope of this change I guess.

xezon commented on Core/Tools/W3DView/AssetInfo.cpp:
```diff
@@ -52,7 +52,7 @@ AssetInfoClass::Initialize ()
 
 		// Assume we are wrapping an instance as apposed to an asset 'name'.
-		RenderObjClass *prender_obj = m_pRenderObj;
+		RenderObjClass *prender_obj = m_pRenderObj.Peek();
```
Use `RefCountPtr<RenderObjClass>` to hold the reference

xezon commented on Core/Tools/W3DView/AssetInfo.h:
```diff
@@ -37,7 +37,7 @@
 
 #include "WW3D2/rendobj.h"
-#include "Utils.h"
+#include "WWLib/ref_ptr.h"
```
After https://github.com/TheSuperHackers/GeneralsGameCode/pull/3059 is merged, you can remove all new ref_ptr.h includes, if WWCommon.h is already included.

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
```diff
@@ -201,30 +198,26 @@ CW3DViewDoc::CleanupResources ()
     }
 
     // Was there a valid scene light?
     if (m_pCSceneLight)
```
Obsolete null test, several times

bobtista replied:
Cleaned up

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
Obsolete test?

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
Obsolete test?

xezon commented on Core/Tools/W3DView/W3DViewDoc.cpp:
It is odd that this one is not cleared. But the original already did that. Maybe is a bug.
